### ISSUE TITLE
Unity analyzer does not work when there are compile errors in VS but not in Unity

### ISSUE THREAD

**Issue by guneyozsan** [microsoft/Microsoft.Unity.Analyzers#153]
Update: Unity analyzer does not run when VS shows assembly compile errors while Unity compiles fine. See comment below: https://github.com/microsoft/Microsoft.Unity.Analyzers/issues/153#issuecomment-789678745

IDE0051 (private member is unused) is not suppressed for Unity messages like Awake, OnDestroy...etc. ~~if there are more than one MonoBehaviour class in the same file.~~

~~Sometimes it is more convenient to write a new MonoBehaviour and then move to its own file when finished (namespace or similar content convenience). The error is not suppressed until the second MonoBehaviour is moved to its own file.~~

- Version of analyzers assembly: Visual Studio 16.8.6, Visual Studio Tools for Unity 4.8.2.0
- Analyzer rule: IDE0051

Steps or code to reproduce the behavior:
1. Create a MonoBehaviour file such as Example.cs.
2. Create another MonoBehaviour class in the same file.
3. Hover the warning sign on Awake or Start '....'
4. See error.
5. Move the second MonoBehaviour to its own file.
6. Error on Awake disappears.

#### Expected behavior
Error should not be reported for Unity messages.

**Comment by sailro** [comment]
I do not repro with the current `main`.

The following test is passing:
```csharp
[Fact]
public async Task UnusedMethodInSeveralTypesSuppressed()
{
	const string test = @"
using UnityEngine;

public class TestScript : MonoBehaviour
{
    private void Start()
    {
    }
}

public class TestScriptCopy : MonoBehaviour
{
    private void Start()
    {
    }
}
";
	await VerifyCSharpDiagnosticAsync(test,
		ExpectSuppressor(MessageSuppressor.MethodRule).WithLocation(6, 18),
		ExpectSuppressor(MessageSuppressor.MethodRule).WithLocation(13, 18)
	);
}
```

Same for a live test in VS (16.8.6):

![image](https://user-images.githubusercontent.com/638167/109530392-836d4d80-7a6b-11eb-8075-721d12cd684e.png)

Reviewing the diff between `main` and your version (tag `1.8.2.0` in this repo), we recently made a [change](https://github.com/microsoft/Microsoft.Unity.Analyzers/compare/1.8.2...1.10.0#diff-7f2b1a265fa4ebbf6e8a4464413a1e8e2f36262b84be3a3898e6272965547046) to better locate all suppressible nodes so I guess it fixed this particular issue.

**Comment by sailro** [comment]
No I do not repro as well with tag `1.8.2`, I am missing something ?

**Comment by sailro** [comment]
Are you sure you didn't confuse the [suppressor](https://github.com/microsoft/Microsoft.Unity.Analyzers/blob/main/doc/USP0003.md) with the fact that we have a dedicated diagnostic for empty [messages](https://github.com/microsoft/Microsoft.Unity.Analyzers/blob/main/doc/UNT0001.md)? 😄

**Comment by guneyozsan** [comment]
Yes I am sure it was not the empty message. But I couldn't repro it after that either.
I'll update if I encounter it again and try to find out the specific case.

**Comment by sailro** [comment]
Got it. Closing this ticket for now but if we manage to find extra repro steps I'll re-open.

Thank you!

**Comment by guneyozsan** [comment]
@sailro I encountered this once again and it has nothing to do with 2 MonoBehaviours being in the same file. Apparently this happens when Visual Studio fails to load an assembly of a plugin used in the project.

In the sample screenshot the warning started to show when I changed the base class from MonoBehaviour to MenuView. Then it started showing the warning everywhere.

Unity compiles and works fine but compiling in VS still fails. It happens pretty randomly and occasionally. Build, rebuild or clean solution or assembly does not work. It fixes itself only when I close and open Visual Studio.

I don't know if this is related with the VS Unity extension but I hope this helps.

IDE0051 on Awake:
![image](https://user-images.githubusercontent.com/19275187/109804979-f2959e00-7c33-11eb-9e65-e9d60904328b.png)
![image](https://user-images.githubusercontent.com/19275187/109805046-05a86e00-7c34-11eb-89cc-cf075b1e1ea6.png)

Assembly error:
![image](https://user-images.githubusercontent.com/19275187/109805140-240e6980-7c34-11eb-8329-3cb9e8c51644.png)
![image](https://user-images.githubusercontent.com/19275187/109805160-296bb400-7c34-11eb-981f-3e1fe862694d.png)

**Comment by guneyozsan** [comment]
Oh, also notice the warning sign (...) on private fields in one of the screenshots as well. Maybe the Unity extension becomes completely inactive.

**Comment by sailro** [comment]
^ @jbevain

**Comment by jbevain** [comment]
@guneyozsan hi!

> Maybe the Unity extension becomes completely inactive.

Something is definitely wrong. What seems to happen is that because the code doesn't compile in VS, our suppressors are somehow not running or are not able to determine that the types are extending `MonoBehaviour`, and thus Visual Studio prompts the default suggestions.

One thing you might try is to disable a Tools for Unity optimization, you can set `Tools -> Options -> Tools for Unity -> Disable Full build of projects` to `False`. You might experience slightly slower builds, but I'm curious to see if that optimization is impacting you.

We'd love to dive more into this issue. Feel free to reach out to me directly, my email is my GitHub alias at microsoft. Thank you!

**Comment by guneyozsan** [comment]
My setup is:
Visual Studio 16.8.6,
Visual Studio Tools for Unity 4.8.2.0
Unity 2020.2.f1
Visual Studio Editor package: 2.0.5 (Unity version mentioned 2020.1.0a12)

I disabled the full build option and see if it has any effect. But this is hard to notice as this happens randomly while switching back and fort VS and Unity.

I'm sending an email as well.

**Comment by sailro** [comment]
Be believe this should be solved with the VS Editor Package `2.0.7`. Please update using the Unity Package Manager.

We'll re-open in case you are still hitting an issue.
